File: src/main/java/com/ptmanager/backend/workplace/WorkplaceController.java
package com.ptmanager.backend.workplace;

import java.util.List;

import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;

import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

import com.ptmanager.backend.domain.Workplace;
import com.ptmanager.backend.labor.LaborCostReport;
import com.ptmanager.backend.labor.LaborCostService;

import java.time.LocalDate;

import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestParam;

@RestController
@RequestMapping("/api/workplaces")
public class WorkplaceController {

    private final WorkplaceService workplaceService;
    private final LaborCostService laborCostService;

    public WorkplaceController(WorkplaceService workplaceService, LaborCostService laborCostService) {
        this.workplaceService = workplaceService;
        this.laborCostService = laborCostService;
    }

    @GetMapping
    public List<Workplace> findWorkplaces() {
        return workplaceService.findWorkplaces();
    }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public Workplace createWorkplace(@Valid @RequestBody CreateWorkplaceRequest request) {
        return workplaceService.createWorkplace(request.name(), request.address());
    }

    @GetMapping("/{id}/labor-cost")
    public LaborCostReport laborCost(
        @PathVariable long id,
        @RequestParam @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate from,
        @RequestParam @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate to
    ) {
        return laborCostService.calculate(id, from, to);
    }

    public record CreateWorkplaceRequest(
        @NotBlank String name,
        String address
    ) {
    }
}

File: src/main/kotlin/com/ptmanager/backend/workplace/WorkplaceController.kt
package com.ptmanager.backend.workplace

import com.ptmanager.backend.domain.User
import com.ptmanager.backend.domain.UserRole
import com.ptmanager.backend.domain.Workplace
import com.ptmanager.backend.workplace.dto.CreateWorkplaceRequest
import com.ptmanager.backend.workplace.dto.QrTokenResponse
import com.ptmanager.backend.workplace.dto.UpdateWageRequest
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PatchMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/workplaces")
class WorkplaceController(
    private val workplaceService: WorkplaceService,
) {

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    @PreAuthorize("hasRole('EMPLOYER')")
    fun createWorkplace(@Valid @RequestBody request: CreateWorkplaceRequest): Workplace =
        workplaceService.createWorkplace(request.name, request.address)

    @GetMapping("/{workplaceId}")
    fun getWorkplace(@PathVariable workplaceId: Long): Workplace =
        workplaceService.getWorkplace(workplaceId)

    @GetMapping("/{workplaceId}/members")
    fun members(
        @PathVariable workplaceId: Long,
        @RequestParam(required = false) role: UserRole?,
    ): List<User> = workplaceService.findMembers(workplaceId, role)

    @GetMapping("/{workplaceId}/qr-token")
    @PreAuthorize("hasRole('EMPLOYER')")
    fun qrToken(@PathVariable workplaceId: Long): QrTokenResponse =
        QrTokenResponse(workplaceService.issueQrToken(workplaceId))

    @PatchMapping("/{workplaceId}/members/{userId}/wage")
    @PreAuthorize("hasRole('EMPLOYER')")
    fun updateMemberWage(
        @PathVariable workplaceId: Long,
        @PathVariable userId: Long,
        @Valid @RequestBody request: UpdateWageRequest,
    ): User = workplaceService.updateMemberWage(workplaceId, userId, request.hourlyWage)
}
